filesystem/filesystem.go

```diff
@@ -393,24 +393,43 @@ func (m *Mount) AddProtector(data *metadata.ProtectorData) error {
 }
 
 // AddLinkedProtector adds a link in this filesystem to the protector metadata
-// in the dest filesystem.
-func (m *Mount) AddLinkedProtector(descriptor string, dest *Mount) error {
+// in the dest filesystem, if one doesn't already exist.  On success, the return
+// value is a nil error and a bool that is true iff the link is newly created.
+func (m *Mount) AddLinkedProtector(descriptor string, dest *Mount) (bool, error) {
 	if err := m.CheckSetup(); err != nil {
-		return err
+		return false, err
 	}
 	// Check that the link is good (descriptor exists, filesystem has UUID).
 	if _, err := dest.GetRegularProtector(descriptor); err != nil {
-		return err
+		return false, err
+	}
+
+	linkPath := m.linkedProtectorPath(descriptor)
+
+	// Check whether the link already exists.
+	existingLink, err := ioutil.ReadFile(linkPath)
+	if err == nil {
+		existingLinkedMnt, err := getMountFromLink(string(existingLink))
+		if err != nil {
+			return false, err
+		}
+		if existingLinkedMnt != dest {
+			return false, errors.Wrapf(ErrFollowLink, "link %q points to %q, but expected %q",
+				linkPath, existingLinkedMnt.Path, dest.Path)
+		}
+		return false, nil
+	}
```

Comment:
I'm trying to figure out when this error condition (the link exists but is "wrong) could actually happen.

Would it make sense to "fix" the protector here?

Reply:
It could happen if someone reinstalls their root filesystem and restores the `.fscrypt` directory.  While I'd like to provide a way to automatically fix protector links, I think that if we were going to do that we should do it properly, maybe providing a `fscrypt fix-protectors` command that would handle fixing all linked protectors on all filesystems.  This particular code is only handling adding a protector to a policy, so it's not really the right place to try to fix the protector.  It won't run if someone is simply trying to use an existing encrypted directory.